api/users/client/client_test.go

```diff
@@ -160,99 +160,334 @@ func TestClient_ListThreads(t *testing.T) {
 		sk, _, err := crypto.GenerateEd25519Key(rand.Reader)
-		require.Nil(t, err)
+		require.NoError(t, err)
 		tok, err := threads.GetToken(ctx, thread.NewLibp2pIdentity(sk))
-		require.Nil(t, err)
+		require.NoError(t, err)
 		ctx = thread.NewTokenContext(ctx, tok)
 		res, err := client.ListThreads(ctx)
-		require.Nil(t, err)
+		require.NoError(t, err)
 		assert.Equal(t, 0, len(res.List))
 
 		// Got one
 		ctx = common.NewThreadNameContext(ctx, "foo")
 		err = threads.NewDB(ctx, thread.NewIDV1(thread.Raw, 32))
-		require.Nil(t, err)
+		require.NoError(t, err)
 		res, err = client.ListThreads(ctx)
-		require.Nil(t, err)
+		require.NoError(t, err)
 		assert.Equal(t, 1, len(res.List))
 		assert.Equal(t, "foo", res.List[0].Name)
 		assert.True(t, res.List[0].IsDB)
 	})
 
 	t.Run("insecure keys", func(t *testing.T) {
 		key, err := hub.CreateKey(common.NewSessionContext(ctx, dev.Session), hubpb.KeyType_ACCOUNT, false)
-		require.Nil(t, err)
+		require.NoError(t, err)
 		ctx := common.NewAPIKeyContext(ctx, key.Key)
 
 		// Got one
 		res, err := client.ListThreads(ctx)
-		require.Nil(t, err)
+		require.NoError(t, err)
 		assert.Equal(t, 1, len(res.List))
 
 		// Got two
 		_, err = net.CreateThread(ctx, thread.NewIDV1(thread.Raw, 32))
-		require.Nil(t, err)
+		require.NoError(t, err)
 		res, err = client.ListThreads(ctx)
-		require.Nil(t, err)
+		require.NoError(t, err)
 		assert.Equal(t, 2, len(res.List))
 		assert.False(t, res.List[1].IsDB)
 	})
 }
 
+func TestClient_SetupMail(t *testing.T) {
+	t.Parallel()
+	conf, client, hub, threads, _, _ := setup(t)
+
+	dev := apitest.Signup(t, hub, conf, apitest.NewUsername(), apitest.NewEmail())
+	key, err := hub.CreateKey(common.NewSessionContext(context.Background(), dev.Session), hubpb.KeyType_USER, false)
+	require.NoError(t, err)
+
+	ctx := common.NewAPIKeyContext(context.Background(), key.Key)
+	sk, _, err := crypto.GenerateEd25519Key(rand.Reader)
+	require.NoError(t, err)
+	from := thread.NewLibp2pIdentity(sk)
+	tok, err := threads.GetToken(ctx, from)
+	require.NoError(t, err)
+	ctx = thread.NewTokenContext(ctx, tok)
+
+	inbox, sentbox, err := client.SetupMail(ctx)
+	require.NoError(t, err)
+	assert.NotEmpty(t, inbox)
+	assert.NotEmpty(t, sentbox)
+
+	// should be idempotent
+	_, _, err = client.SetupMail(ctx)
+	require.NoError(t, err)
+}
+
+func TestClient_SendMessage(t *testing.T) {
+	t.Parallel()
+	conf, client, hub, threads, _, _ := setup(t)
+
+	dev := apitest.Signup(t, hub, conf, apitest.NewUsername(), apitest.NewEmail())
+	key, err := hub.CreateKey(common.NewSessionContext(context.Background(), dev.Session), hubpb.KeyType_USER, false)
+	require.NoError(t, err)
+
+	from, fctx := setupUserMail(t, client, threads, key.Key)
+	to, _ := setupUserMail(t, client, threads, key.Key)
+
+	res, err := client.SendMessage(fctx, from, to.GetPublic(), []byte("howdy"))
+	require.NoError(t, err)
+	assert.NotEmpty(t, res.ID)
+	assert.NotEmpty(t, res.CreatedAt)
+}
+
+func TestClient_ListInboxMessages(t *testing.T) {
```

Comment:
Here's a nice list message test.
Message are by default listed in descending order _by ID_. Since ULIDs are based on time, we get time ordering for free, plus efficient pagination by telling the datastore to seek keys.
Depends on https://github.com/textileio/go-threads/pull/411.

Reply:
Ah yeah. Good point. For now, this is a pure-hub thing since it's connecting hub users.